sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/pointer/CommonPointerLibraries.java

```diff
@@ -116,6 +138,48 @@ static boolean isMemberModifiable(LLVMPointerImpl receiver, String ident,
         }
     }
 
+    @ExportMessage
+    static boolean isMemberInvocable(LLVMPointerImpl receiver, String ident) {
+        LLVMInteropType type = receiver.getExportType();
+        if (type instanceof LLVMInteropType.Clazz) {
+            LLVMInteropType.Clazz clazz = (LLVMInteropType.Clazz) type;
+            return clazz.findMethod(ident) != null;
+        }
+        return false;
+    }
+
+    @ExportMessage
+    static Object invokeMember(LLVMPointerImpl receiver, String member, Object[] arguments)
+                    throws UnsupportedMessageException, ArityException, UnknownIdentifierException, UnsupportedTypeException {
+        LLVMInteropType type = receiver.getExportType();
+        if (!(type instanceof LLVMInteropType.Clazz)) {
+            throw UnsupportedTypeException.create(new Object[]{receiver}, receiver + " cannot be casted to LLVMInteropType.Clazz");
+        }
+        // change from receiver.foo(arguments) to interopLibrary.execute(foo, [receiver+arguments])
+        Object[] newArguments = new Object[arguments.length + 1];
+        newArguments[0] = receiver;
+        for (int i = 0; i < arguments.length; i++) {
+            newArguments[i + 1] = arguments[i];
+        }
+        LLVMInteropType.Clazz clazz = (LLVMInteropType.Clazz) receiver.getExportType();
+
+        Method method = clazz.findMethod(member, newArguments);
+        if (method == null) {
+            throw UnknownIdentifierException.create(member);
+        }
+        LLVMFunction llvmFunction = LLVMLanguage.getContext().getGlobalScope().getFunction(method.getLinkageName());
+        if (llvmFunction == null) {
+            CompilerDirectives.transferToInterpreter();
+            final String clazzName = clazz.toString().startsWith("class ") ? clazz.toString().substring(6) : clazz.toString();
+            final String msg = String.format("No implementation of declared method %s::%s (%s) found", clazzName, method.getName(), method.getLinkageName());
+            throw new LLVMLinkerException(msg);
+        }
+
+        LLVMFunctionDescriptor fn = LLVMLanguage.getContext().createFunctionDescriptor(llvmFunction);
+
+        return InteropLibrary.getUncached().execute(fn, newArguments);
```

Comment:
You can use `@CachedLibrary InteropLibrary interop` to get a more efficient version of the interop library.